Adds subscription apis and validate version support

venkhatv opened the pull request:
Adds subscription apis and validate version support

ashvinikumars commented on lib/consumers.js:
```diff
@@ -109,6 +109,79 @@ ConsumerInstance.prototype.subscribe = function(topic, options) {
+    }.bind(this));
+}
+
+ConsumerInstance.prototype.getSubscriptionDetails = function(res) {
```
We could also do it like this,
```
ConsumerInstance.prototype.getSubscriptionDetails = function(res) {
    return new Promise((resolve, reject) => {
        this.client.request(utils.urlJoin(this.getUri(), 'subscription'), (err, result) => {
            if (res) {
                res(err, result);
            }
            if (err) {
                reject(err);
            } else {
                resolve(result);
            }
        });
    });
}
```

ashvinikumars commented on examples/console_consumer.v2.js:
```diff
@@ -0,0 +1,201 @@
+    argv = require('minimist')(process.argv.slice(2));
+
+    argv = {
+        url: "https://skiff-api.dev.saas.nutanix.com",
```
Use generic urls like http://localhost:8082 instead of company specific ones. Same for group and topic. test-group and test-topic.

venkhatv commented:
1. Added support for all new consumer apis
2. Added promise support, version check for v2 and v1 apis in consumers n brokers.

Things to do:
1.  Add Promise and version check in Topics, partitions.
2. Test throughly v1, v2 apis in All files, via callback and promise .
3.  Unit test cases
4. Clean up example files

venkhatv commented:
Cleaned up - unneccesary code, comment from all files

Things to do:
Write unit test cases

venkhatv commented:
1. Added Unit test using Mocha, Chai, Sinon
2. Formatted lib files, examples, test files.

Pls check formatting, indendation, and code readability

ashvinikumars commented on examples/console_consumer.v2.js:
```diff
@@ -0,0 +1,126 @@
+    requestDelay: 1000 * 10
+};
+if (fromBeginning) {
+    consumerConfig['auto.offset.reset'] = 'smallest';
```
This won't work. Valid values are earliest, latest and none. The value influences the initial offset when there is no saved offset for the consumer group. This would happen when the consumer group is connecting for the first time or when the saved offset is out of range. If the user specifies a consumer group name and wants to re-run the example a second time, consuming messages from beginning this will not work. There are APIs available to move the offset to beginning before starting to consume messages. These APIs need to be used. 
To keep the example simple, either remove the option to specify consumer group name so that we always create a new consumer group or remove the fromBeginning.

ashvinikumars commented on examples/console_consumer.v2.js:
```diff
@@ -0,0 +1,126 @@
+    "auto.commit.enable": "true"
+};
+/** Update this to add more topics to subscirption*/
+var subscriptionsTopics = {
```
Rename variable to subscribedTopics

ashvinikumars commented on lib/client.js:
```diff
@@ -142,6 +142,20 @@ Client.prototype.request = function(opts, res) {
     req.on('error', this._handleRequestError.bind(this, res));
 }
 
+//Use this, when body is sent in Get request
```
Which REST proxy API requires/ allows a body in GET method?

ashvinikumars replied:
Just noticed that the Get committed offsets API accepts a request body.

ashvinikumars commented on examples/console_consumer.v2.js:
```diff
@@ -0,0 +1,126 @@
+if (help || topicName === undefined || (format != "binary" && format != "avro")) {
+    console.log("Consumes and prints values of messages from a Kafka topic via the REST proxy API wrapper.");
+    console.log();
+    console.log("Usage: node console_consumer.js [--url <api-base-url>] --topic <topic> [--group <consumer-group-name>] [--message-limit <num_messages>] [--from-beginning] [--format <avro|binary>]");
```
Limit line length to 120 to avoid need of horizontal scrolling when viewing in Github
